### PR TITLE
Fix change detection in audioDevices list

### PR DESCRIPTION
**Pull request by 2yangk23**
This seems to fix the double init, as well as https://github.com/ppy/osu-framework/issues/3377

Let me know if you find any problems.

### CODE REVIEW

File: osu.Framework.Tests/Audio/AudioThreadTest.cs [hunk 1]
```diff
@@ -37,6 +38,8 @@ public void TearDown()
             thread.Exit();
 
             WaitForOrAssert(() => thread.Exited, "Audio thread did not exit in time");
+
+            Bass.Free();
```
Comment (peppy) [inline osu.Framework.Tests/Audio/AudioThreadTest.cs]: why is this required if it's happening in the thread shutdown logic?

image

File: osu.Framework/Audio/AudioManager.cs [hunk 2]
```diff
@@ -77,6 +77,8 @@ public class AudioManager : AudioCollectionManager<AdjustableAudioComponent>
             MaxValue = 1
         };
 
+        public override bool IsLoaded => Bass.CurrentDevice != Bass.DefaultDevice;
```
Comment (peppy) [inline osu.Framework/Audio/AudioManager.cs]: Is there not a case where `CurrentDevice` can be `DefaultDevice`?

Comment (2yangk23) [reply osu.Framework/Audio/AudioManager.cs]: From documentation of CurrentDevice:
> Gets or sets the device setting of the current thread... 0 = no sound, 1 = first real output device.
> A return value of -1 indicates error.

DefaultDevice = -1, but I could use -1 here instead if you think that's more clear. You can't actually initialize DefaultDevice so you could consider it invalid.

File: osu.Framework.Tests/Audio/AudioThreadTest.cs (hunk 1 above)
Comment (2yangk23) [reply osu.Framework.Tests/Audio/AudioThreadTest.cs]: You're right it is calling Free(), but I guess Free() does not reset CurrentDevice back to -1. This causes the next test to be unable to initialize properly and it gets stuck waiting on preconditions. For whatever reason, calling Free() twice set CurrentDevice back to -1, but I'll find a better way to solve this.

File: osu.Framework/Audio/AudioManager.cs (hunk 2 above)
Comment (2yangk23) [reply osu.Framework/Audio/AudioManager.cs]: Very strange behavior:
>[AudioThread] Free() device:0
[AudioThread] result:True error:OK device:2
[DeviceLosingAudioTest] Free() device:2
[DeviceLosingAudioTest] result:True error:OK device:-1

Comment (peppy) [reply osu.Framework/Audio/AudioManager.cs]: Might be worth isolating this to ensure nothing is calling `Init` or another function between the two free calls

File: osu.Framework.Tests/Audio/AudioThreadTest.cs (hunk 1 above)
Comment (peppy) [reply osu.Framework.Tests/Audio/AudioThreadTest.cs]: Might be worth isolating this to ensure nothing is calling Init or another function between the two free calls

File: osu.Framework/Audio/AudioManager.cs (hunk 2 above)
Comment (2yangk23) [reply osu.Framework/Audio/AudioManager.cs]: Actually I think I understand it now. Each time you call `Free()` it just frees a single device. But it is possible that we have multiple devices initialized and you need to keep calling `Free()` until `CurrentDevice` is back to `DefaultDevice`. Note that this is observed behavior and I'm not sure why it isn't documented...

File: osu.Framework.Tests/Audio/AudioThreadTest.cs (hunk 1 above)
Comment (2yangk23) [reply osu.Framework.Tests/Audio/AudioThreadTest.cs]: Actually I think I understand it now. Each time you call Free() it just frees a single device. But it is possible that we have multiple devices initialized and you need to keep calling Free() until CurrentDevice is back to DefaultDevice. Note that this is observed behavior and I'm not sure why it isn't documented...

Comment (peppy) [reply osu.Framework.Tests/Audio/AudioThreadTest.cs]: http://www.un4seen.com/forum/?topic=16664.msg116540#msg116540 this is the only reference i can find, which suggests that the above may only happen if we are calling `Init` multiple times.

Comment (2yangk23) [reply osu.Framework.Tests/Audio/AudioThreadTest.cs]: Yea we are calling `Init` multiples times. For example the test `TestPlaybackWithDeviceLoss` starts by initializing some device, the "loses" that device, which causes the NoSound device to get initialized. (2 devices are now initialized). When the audio thread was exiting, `Free` was only called once which only frees one of the devices. Then when the next test started, there was already an initialized device which caused problems.

Comment (peppy) [reply osu.Framework.Tests/Audio/AudioThreadTest.cs]: I see.

So in teardown, let's do a `while` over freeing i guess. add a comment inline mentioning why we are doing this.

**Comment by peppy** [comment]
@2yangk23 sorry for the back and forth, but could you please check my latest commit? I think it should resolve the issues without resorting to a `while` loop.

**Comment by 2yangk23** [comment]
Functionally it's working fine, but the test is still broken so you may not be calling Free() for all devices.

**Comment by peppy** [comment]
Will take another look into this today (I can't repro the failure on a macOS host, seems windows specific).

**Comment by 2yangk23** [comment]
I looked at your code and it looks like you added the `Free` call when we get `Errors.Already`. The problem is you only get `Errors.Already` when you initialize the same device twice, not if you initialize two different devices. The problem that causes the test to fail is when you initialize two different devices at the same time (e.g. NoSound device + Real device).

**Comment by peppy** [comment]
Right, I see. That makes sense as to why it's still failing then. Il'l add back the loop on top of my changes with a comment. Maybe we'll be able to resolve it better one day.
